### ISSUE TITLE
Provide optional syncing of client settings.

### ISSUE THREAD

**Issue by creesch** [thelounge/thelounge#1825]
Issue made after discussing it on irc. The idea is to implement client settings sync based on the following principles. 

1. Settings sync is an *optional* setting per client. Explicitly done like this to prevent people loosing their current settings. 
2. All settings will be synced. More fine grained settings sync will be for a follow up. The idea is to start simple and based on the fact that there currently is no sync so this is already an improvement. Previous attempts stranded on trying to go for fine grained from the get go. 

The work flow for this implementation will be roughly something like this. For clients that have settings sync enabled: 

- On client load settings are picked up from the server and applied. (this requires reading the current client settings to see if this is the case.)
- When a setting is changed in a client the settings object will be saved to the server. 
- The server pushes the changed settings to other sync enabled clients that are currently active. To keep things simple settings will not be applied on the fly but the user will be presented with a message giving them the option to reload the client or disable syncing for that client. Something roughly like this 
![image](https://user-images.githubusercontent.com/2502190/33842794-dbe5180e-de9b-11e7-9dd4-980f9209e97a.png)

**Comment by creesch** [comment]
Previous attempt https://github.com/thelounge/lounge/compare/settings

**Comment by xPaw** [comment]
Important to note specific settings like "custom highlights" would have to be always synced to provide correct support for push notifications. Right now these custom highlights are done entirely on the client and not the server.

**Comment by astorije** [comment]
Hey @creesch, thanks for taking that up!

I think however this should be handled in a different way: at least for now, there should be no UI change introduced by this whatsoever.
In general, I don't think the user should be prompted for settings sync. Heck, I'm trying to think really hard on all the apps I use on multiple devices that have settings, and of the dozen-or-so I daily use, I have never seen something like this.

So anyway, this is how I think we should handle this:

- Some settings like the custom highlights, should be always synced. Like @xPaw said, we currently do that on the client, and it's hurting: push notifications do not have a way of knowing what they are.
  There are other things we currently do not handle correctly because of lack of sync, such as the auto-away message: it is currently stored in the user's JSON configuration file. Not user-friendly at all :D This is also probably true for the `commands` array we store in user config

- Most client-only settings actually do not need settings sync, such as "Show seconds in timestamps", for several reasons: you might want to enable that only on big screens, and "replicating" this setting is _just_ selecting a checkbox, not typing long strings across 12 devices. With reasonable defaults, we make it work for 80% use cases, and the other 20% will enable/disable the checkbox as they please.
  This is true for the theme as well: extremely easy to pick one from your phone, and it might actually be reasonable to want to use a different theme on mobile if that other theme is much better for mobile. Simple enough, and no need to have complex code supporting replicating a theme to 2-5 devices (which is probably the average or a little more)

- Some settings are not or hardly impacted by this. In our case, the password reset fields, and the push notification subscription button (at least in its current state).

- Some one-of-a-kind settings will require more thinking, probably ad-hoc. I'm thinking mainly about the custom CSS setting where, indeed, it might be interesting to sync after you set dozens of lines that match your needs. However, should you use 2 different themes or have different expectations depending on devices, you would really not want them to always be in sync.
  In these snowflake cases, it is better to make "tailored" decisions. For example, for the custom CSS, we could have a checkbox that says "Use this CSS on all devices", with a source-of-truth CSS for all devices except ones without this checkbox checked, etc. (this is just an example ;) )

**But** what the first and last items have in common is that they would absolutely leverage a common way to sync some settings. So I think the first step to achieve this would be to make it possible, in the code, to make some options "syncable". A good first candidate that could be a first, nice and simple (well, relatively) PR would be custom highlights. That would help make a proof-of-concept of settings to be synced without having to think a whole UI/UX.

A second step could be to bring what we already have but can't use on the UI yet, like [auto-away message](https://github.com/thelounge/lounge/blob/2d0ddfb2e84054722bce97c703aa86f553e5ddc5/src/client.js#L74).

Then we need to ask ourselves what do we want to always sync and not sync. And we can then tackle cases like the custom CSS box which is its own thing.

Sorry for the loooong post. I hope that was clear enough, feel free to highlight parts that need clarifying 🙂

**Comment by creesch** [comment]
Sorry, I should have clarified it a bit more. This issue is a result of a discussion on irc and how I described it is sort of the basis design principle.

Everything you described is something that could eventually be implemented. The first goal is to have sync period, everything else is a bonus.

**Comment by astorije** [comment]
> Everything you described is something that could eventually be implemented. The first goal is to have sync period, everything else is a bonus.

Then I'm a bit confused, because in my mind any UI behavior comes on top of a syncing mechanism like what I described above.
I guess my overall comment would be that we should absolutely not have a setting to enable syncing settings (it's rather meta lol).

**Comment by creesch** [comment]
If we don't have that setting peoples current settings on various clients might get missed up. It is something that consistently came up in the discussion on irc.

That discussion is summed up in the issue and that is what I am currently building.

**Comment by astorije** [comment]
Again, sorry if I'm not being clear lol. It's not so much about the setting itself but the UX issue that it's hiding.

We should not sync all settings across devices. This represents poor UX on our part, and a lack of making adequate decisions for individual settings. Also, it is going to be a massive pain to maintain, fix, improve, and get away from when we want to do something more atomic.

Instead, it should be possible to tell which settings we sync with the server. For what you are building, if you just make "custom highlights" a setting synced with the server, that would be a terrific feature and set the ground for more and better settings improvements. That is something we desperately need :)

**Comment by creesch** [comment]
And we are not going to do it on a per setting basis but on a per client basis. This provides a much simpler approach to this all.

What you are doing is exactly the same as a bunch of people on irc saying "but X is most important" and is imho one reason previous attempts tried to do too much.

Just to give an example, for me the various css improvements I have to keep in line are more important while I don't want the same highlights in all clients.

Which is why we decided to keep it extremely simple and sync all settings across clients **if** the user has applied one checkbox in those clients that opts them in.

The goal is to have sync first in a way that effectively works for everyone and can be refined later.

I invite you to read the scrollback in irc because this honestly is something we discussed rather thoroughly.

**Comment by creesch** [comment]
https://github.com/thelounge/lounge/compare/master...creesch:settingSync

Building it in here, the basic principle already works rather well.

**Comment by astorije** [comment]
I did read the scrollback and I don't agree with this issue, sorry.

> And we are not going to do it on a per setting basis but on a per client basis.

Why? This is literally what every single app out there does. What we're about to do is, in my experience, unheard of. I'm sorry if that sounds harsh but that would be a UX fail.
What you find easier to do now is going to be a maintenance/evolution hell for us in the near future, because when we decide to improve this, we'll have to deal with this legacy, remove extraneous data (instead of adding necessary ones), and [there will be riots](https://xkcd.com/1172/) :D

> Just to give an example, for me the various css improvements I have to keep in line are more important [...]

I completely understand that you'd want synced custom CSS. In fact, when we introduced that custom CSS feature, that's something that came up in the discussion: what if I want to sync my custom CSS across my devices? We did not flag that as a priority for the feature, but I absolutely understand the need.

This is why I'm saying step 1 should be as simple as making one option server-sync. I suggested custom highlights because it's a simple "always sync" situation, but if what you really care about is custom CSS, then by all means, let's add a way to sync custom CSS by checking a checkbox under the textarea that reads "Sync this snippet across all my devices" or whatnot. That will be a great addition.
But syncing _all_ settings, no no.

> [...] while I don't want the same highlights in all clients.

As @xPaw said, this is not going to be an option. Custom highlights **need** to be on the server side of things. What we have right now has a few bugs that will live as long as we keep this client-only. We've known for a while that we would need to keep this on the server eventually :)

**Comment by creesch** [comment]
If the way I am currently building it is going to be a deal breaker, then someone else will have to build it. I have seen the previous attempts and they mostly stranded on people arguing over what should or shouldn't be included.

Furthermore, I don't think you can at one point say you want sync to be mandatory and then link that xkcd with a straight face. People have different configurations for their mobile and desktop clients for example and are not going to like it one bit if one overwrites the other. And again, this is something discussed on irc and you are in the minority here as it is the foremost concern brought up.
You cannot project your own preferences as UX truth, certainly when putting it in hyperbolic statements saying it will be a complete fail. 

I proposed a simple solution for that yesterday which was well received and based on the following discussion made this issue to sum up the discussion and have a basis to work from. It also isn't one that is unmanageable at all in the future (again hyperbolic), it is a very simple KISS basis that allows further expansion. 

I honestly don't have the energy to do the same discussion in here a bunch of times again and again.

So again, if it being built as outlined in the easy is going to be a deal breaker I am out. 

@xPaw @YaManicKill

**Comment by astorije** [comment]
> I don't think you can at one point say you want sync to be mandatory

I don't recall having said that, because I do not want that. I kept rooting for the opposite of that in this thread. The only thing you could see as me saying is "mandatory" is the custom highlights part. I'm not saying I want this to be mandatory, I'm saying it needs to be synced to fix existing issues.
I do not want to force users to sync settings. In fact, I want them to be able to sync custom CSS without having to sync everything else (as mentioned above).

> and then link that xkcd with a straight face

Don't take things that way, it was an innocuous link, not an insult ;)

> People have different configurations for their mobile and desktop clients for example and are not going to like it one bit if one overwrites the other.

I feel like we misunderstand each other: I **do not** want to force synced settings across all devices. I have never argued in favor of that either in this channel or ever :)

> You cannot project your own preferences as UX truth

I do not. This is not my own preferences, this is my experience after working and maintaining this project for well over 2 years, and working hand-in-hand with a UX team daily as part of my day job. It is not hyperbolic: I am not aware of a single app that requires user intervention and syncs their entire settings across devices.
And yes, I do believe that the direction being taken is not the right one. It's my opinion, not my preference.

The maintainability issue I am trying to bring up is this: right now users cannot sync anything (bad). Then users would be able to sync everything or nothing (also bad). When we'll try to do the right thing, i.e. only sync what needs to be sync, we'll have to tell users that, sorry, now you can't sync this or that setting. It did not make sense to do it, but because we enabled it, you can be sure some users will rely on it and be upset (hence mandatory xkcd strip).

What I'm not sure to understand is: if what you really care about is to sync custom CSS across all devices, why not doing that? I think we all agree on that (while, again, syncing everything is the problem) so that would be a good start.

**Comment by creesch** [comment]
Irccloud simply syncs all settings across all clients, when we discussed that in irc it was seen as a good option but not in a mandatory fashion as that would wipe out so we offer the option "sync this specific client". It is a simple, easy to understand mechanism. 

The way I am seeing it you are now lobbying for a more complex confusing situation where some settings are synced, others are not and where if you want to give control to users you end up with a ton of extra checkboxes (one for each setting to allow syncing them). 

The most simple, user friendly (as in predictable and understandable) way of doing this is to actually sync all settings if the user so wishes. Anything else is more complicated, harder to understand and therefore worse as far as UX goes. 

Custom css would be a high priority for me yeah, but so are other settings while other people have different priority as far as settings go. I named it as an **example** of the many different preferences people have.

Also if you are worried about denying people sync on certain things in the future why not approach it from the other side. Are there any settings that absolutely never should be synced, those can be excluded and the rest we sync.

**Comment by astorije** [comment]
> Are there any settings that absolutely never should be synced, those can be excluded and the rest we sync.

I argue that none should be unless they technically have to (custom highlights, auto away message) or there is a real pain point in not doing so (custom CSS). All other instantaneous-to-replicate settings like checkboxes can be checked on a device-by-device basis.

This is not a foreign idea, this is what pretty much all apps running on phones do today. And with clear presentation it does not have to be confusing, it is much more intuitive.

**Comment by creesch** [comment]
It is honestly not how most apps I know work, it is how I'd say 50% of the apps work. I should note that I am not looking at this is a phone app but a cloudlike webapplications and in that area it is much more common to have settings synced regardless of where you are logged in. 

Which honestly makes more sense to me, your settings are available wherever you are logged in. Meaning you can easily use the lounge on a friends computer, a new computer, work computers (plenty of people have their browser settings reset at their office).

 I personally find apps that do not do this annoying, often enough I have been using a cross device app where it did something that made me wonder "Didn't I change this in settings?" only to discover that settings did not sync. 

The way you are proposing it really doesn't make it simpler it will only make people go "But why can't I sync X if it is technically possible?!". Frankly, it comes across more or less like *"These things I find important and other settings... meh"*. 

And honestly, so far you haven't really given a reason why other settings shouldn't be synced. You just have pointed out that you think some are more important. I do understand custom highlights being important if it breaks things (does it? and why hasn't this been addressed in some way, seems like a serious bug) but other that is just an argument to sync one setting not to keep the others out. 

Also, once more what I am working on is a *starting point* not an end point. 

Anyway, now I have spend my entire commute discussing this... I have made some good progress yesterday, I'll probably continue on it a bit more and then put in a pull request and then ya'll can decide what to do further with it. I will keep what you have said in mind as far as custom highlights goes.

**Comment by astorije** [comment]
You have some valid points regarding use cases that I'd be happy to discuss. It's 3am here so I'm gonna try to get some sleep though, but I'll reply tomorrow.

Oh, just before closing, I did not say that some settings are more important than others :)
(Also yes we've known custom highlights do not work on push notifications, just did not fix it yet, and to the question "why hasn't this been addressed yet" the OSS answer is usually "because I sleep between 3am and 8am", sadly 😁)

**Comment by creesch** [comment]
Everyone has to sleep and I certainly do get the "time available vs amount of things you want to do" thing.

**Comment by creesch** [comment]
A checklist of things that probably need to be done anyway: 

 - [ ] Rewrite settings in such a way that only when things are truly changed it will be written to settings. 
 - [ ] Make it so that individual setting changes are reported to the back end (future proofing)
 - [ ] Only emit changed settings back to clients (again future proofing).

**Comment by YaManicKill** [comment]
@astorije I think I'm a little confused as to what you think we should be doing.

I totally understand your points about users not having to think about this, and for most cases it is absolutely correct that most users will be happy that everything syncs because that's what they expect.

But there are a minority of users who will be vocally against this. We need to allow them a way to not, if they don't want to.

Regarding custom css: my plugin support for custom css will essentially remove most of the functionality of the custom css box, and will add the custom css to all clients. Not massively important to the discussion, but thought I would mention it.

But my question is...you seem to be jumping between different points of view.

Are you saying that settings should all be synced automatically without the user's input or permission? Or are you saying that only some of the settings should be synced automatically without the user's input or permission and the others not synced? Or are you saying that only some of the settings should be synced automatically without the user's input or permission and the others with an option to sync? Or are you saying that the user should be able to choose exactly which settings sync? Or are you saying that a user should be able to choose whether the settings sync, but when they do, all the settings should sync?

I, personally, think we should (at least to start with) sync all of the settings automatically, but give users the ability to turn it off on a specific client if they wish. We need to think about how this works for users that already exist (because we don't want to accidentally lose all their settings just now), and maybe it should be opt-in for users that already exist, but automatic for new users.

**Comment by dgw** [comment]
I'm not in a position to write a reply of any significant length, but I will leave my biggest-picture thought: What if defaults synced from the server, and individual clients overrode them on a setting-by-setting basis locally only for items that differ? It's probably an even bigger design nightmare, but it could be a good compromise. (Obviously custom highlights will be handled on the server eventually and needn't enter into this equation!)

**Comment by astorije** [comment]
Alriiiight so: I was having a hard time getting why my position was confusing. After all, I had on this thread a single position (until now), and I thought I was clear enough. Heck, I even said "But syncing _all_ settings, no no." Clear enough, right? ... Apparently no 😁

So anyway, re-reading all this, it struck me that... we are not talking about the exact same thing. 🎆 It feels to me that this is the root of the whole issue. And from that, hard to communicate. So I totally apologize if I misunderstood the original intent or miscommunicated my views. I'll try to do better in this post.

Server-sync of client settings in our situation mean (at least) 2 things:

- Solving the problems we've been having for a while, where we keep settings on the client that actually totally belong to the server. This is the case for example of the custom highlights, because handling them should be done on the backend, but because we don't have any setting sync, we kept them on the client, and therefore the feature was not fully working. Solving this means a way to make some settings live in the user configuration on the server (exactly what we do for auto-away message for example) but still edit them on the client, next to settings that have no need to reside on the server (for example, enabling notification sounds, it is not _technically_ needed on the server)

- Giving a way to users to save their client settings to the server, kind of the in-app equivalent of a Dropbox for settings.

The key difference is that the first version is a technical requirement, a problematic situation we have to address to solve some of the problems that we already have and that we will have. _Already_ because there is currently no way to set the auto-away message on the client, and _will_ because #1161, #464 (probably), #462, #179 (probably), ... I'm sure there are 20 open issues that need that :D
The second version, however, is not directly aimed at fixing things or enabling other features to work better: sync is not a means to an end like previously described, it is the end itself. It is mainly and mostly a convenience feature for the user: technically speaking, there is no _need_ to store themes or notification sounds on the server, but users might want to.

Anyway, this whole time, my understanding was that you were trying to solve the first issue, and to do so were just syncing everything because why not, and then making it optional. That would have been an issue: the reason to store highlights on the server is specific to that setting, and because of this, not all settings are equal.
This is what showed me that we were not talking about the same thing:

> (plenty of people have their browser settings reset at their office)

This is a use case that _requires_ server sync, while I was trying to find ways to solve underlying problems without server sync. In short, I didn't understand that you were needing this feature itself, which is why the whole time I was trying to find ways **not** to do client setting sync at all.

Okay, I hope this is clearer now. @YaManicKill, your list of questions is still valid, but I can't answer it with what I had in mind the other day because, again, I was not focused on the sync feature itself.

So now that this is cleared up, my thoughts on this feature:

- I think it would be valuable to acknowledge both bullet points I detailed above as separate things. What I mean by this is that trying to address issues like the custom highlights settings (and others that would belong to the same category) with the same tool that your server sync feature offers might not be best.
  If I understand your implementation correctly, you have a way to always sync some client settings (highlights only so far but extensible) and then make sure that other clients are updated when that client-side value change.
  It seems to me that such settings actually belong to the server (the user config on the server to be precise) and should not be client-based at all (i.e. not in the `options` localStorage object. Loading the client or the settings page fetches and displays those settings, and if I update them, they get updated on the server, other connected clients pick up the new value. While this version still uses some parts of your feature (storing on the server, getting/setting from the client), it is decoupled from the all-setting-sync feature.
  I feel like I'm starting to be unclear again, does that make any sense? If not, maybe it will be easier to point out once your PR can be commented on.

- Ultimately, it would be great if solving the issues of client-settings-that-actually-belong-to-the-server (like highlights) would not be completely dependent of the server sync feature. The reason why I'm saying that is because this feature reminds me https://atom.io/packages/sync-settings a lot. Comes the day we are able to package things correctly, "sync-settings" would be a great, self-contained package, but if a user decides to completely disable or remove the package, settings should still work as expected (and in the case of highlights, expected means stored on the server). What brought me to that is [this line](https://github.com/thelounge/lounge/compare/master...creesch:settingSync#diff-03ed49e301a4e588ec406bc5aace2a02R39). I think we should be able to make highlights and similar behave correctly without having to whitelist/blacklist them: in the end, they shouldn't be client settings.

- I understand both sides of the argument, but if I were to pick I would disable server-sync by default. It seems to me (this is not data, it's observation, thus "seems" 🙂) that the use cases that settings sync tries to address are a bit niche. I've mentioned that before, but most apps that I personally use do not sync settings by default unless it is necessary or their main design. Other apps that I know offer that (like Atom and that third-party package) offer to opt-in, not to opt-out. I understand if you have a different experience, this is just my 2c. I would prefer to keep it opt-in for TL, but if you have strong preferences for opt-out instead, either way it's a simple checkbox away. We'll quickly find out if people start opening issues "why tf are my settings not kept on my phone when I set them on my laptop", or "I cannot live without server sync, please make it default", or if no one cares either way lol.

- When/If this feature becomes a package, then it is not unreasonable at this point to offer crazy customization for it. As an example, while Atom's base settings are pretty simple, and some packages will be opinionated and offer little customization that work for 90% of the users, it is not uncommon to find other packages with an insane amount of control. If people need/want/crave for a fine granularity, then sure, but until then, I totally agree that a all-or-nothing approach is an ideal first step. It offers the feature without cluttering the window or making it hard to understand and use.

Alright, I'm sure I had other things to say but it's now almost 2am and I'm sure I already wrote way more than what anyone's patience allows. Damn, that post is almost a design document.
Hope that this is clearer and more in line with what you had in mind. I'm pretty sure there will be a bit of back-and-forth to finalize the implementation (mainly that highlights-and-such should not be stored on the client at all), but after that, the feature sounds logical.

**Comment by YaManicKill** [comment]
So, it sounds to me like there are 2 different kinds of settings here.

Server settings, that don't make sense *not* to be synced (custom highlights are probably the only one currently, but there might be more...there are definitely ones that are only currently on the server and we could put in the client if we had syncing), and client settings that make sense to either be synced or not.

So, I would suggest that we split up our settings page now. We have been talking about doing it before, and @astorije you created some mockups, but I feel like this is the right time.

Then the "server" settings should always be synced, and the "client" settings can be on or off. 

Does that sound good?

Now let's get into the real bikeshedding.

> I've mentioned that before, but most apps that I personally use do not sync settings by default unless it is necessary or their main design. 

Right, this is where (I think) you are thinking of the lounge incorrectly. Don't compare the lounge to standard irc clients, or a text editor. It's not a traditional client. It's a web app. Compare it to other web apps.

So, let's take...facebook as an example. Do you ever have settings that are only on 1 web browser for facebook? That would be really weird and confusing.

But yeah, we aren't like facebook, we are a chat client. So let's take slack, or discord, or facebook messenger. They generally **don't** have client settings that are specific to the browser you use.

Therefore I would argue that actually... **not** syncing all the settings is niche.

But really, I won't argue too much about whether the default for new users should be on or off. As long as we have the feature, that's the main thing. But I think it's confusing to new users to have it turned off by default.

**Comment by creesch** [comment]
I am going with @YaManicKill's thing here. It seems to me that most of what you address in the first points should be the focus of a different PR altogether as this issue is focussed on user settings. Heck, before I started on this I had no idea that there were settings that "do not belong" in that regard. I am getting the feeling that you are still compounding to many issues into this one. Which is also why I'd argue that syncing of settings has failed many times before.

As far as you comparing lounge to locally installed apps, that is as @YaManicKill has pointed out not a fair comparission. Lounge is a web app, a (private) cloud based app and the majority of apps in that category do sync most of their settings. 
You argue for it to be off by default, which is what the intend was from the start. People can turn it on when they like. 

So here is what I propose: 

- I finish this feature based on the settings currently in here and put in a PR.
- A new issue is made to discuss how and if it is needed to refactor settings to be able to separate settings from those that are server centrally dictated and those that can differ on a per client basis.

**Comment by YaManicKill** [comment]
> You argue for it to be off by default, which is what the intend was from the start. People can turn it on when they like.

Yeah, but I'm arguing that it should be on by default for new users.

I think that existing users should have it off by default, but new users should have it on by default.

**Comment by creesch** [comment]
Oh I agree on the new users part, but not to the degree to make that a breaking point. If astorije finds that specifically important I don't mind the concession there. The option to sync is more important to me than the default setting for it.

**Comment by astorije** [comment]
> So, it sounds to me like there are 2 different kinds of settings here.

Exactly. And when I look at settings of Telegram or Slack, I realize that it's the same for them both!

> So, I would suggest that we split up our settings page now. We have been talking about doing it before, and @astorije you created some mockups, but I feel like this is the right time.
> Then the "server" settings should always be synced, and the "client" settings can be on or off.

I'm totally in favor of splitting settings, but I would strongly recommend not splitting them client/server.
It's anecdotical, but I had a chat not long ago with a friend UX designer (I tried to grab him for TL, but that didn't work :( ) who was telling me that the engineers in his team had designed app settings in a way that highlighted the server capabilities, and their support team was constantly dealing with users misunderstanding how to use them :)

If I may suggest something (though I don't want to derail the focus on this issue), I would recommend to split them logically not in terms of server capabilities but in terms of what they're actually used for. For example (don't take this at face value, it's just to illustrate and there are probably better tweaks), we could imagine a "Notifications" tab that contains our current browser/push notification checkbox + sound + custom highlights, and in this case, only the last one is server-synced. How to differentiate their behavior then? I think that a nice/short user-friendly statement underneath to clarify "Custom highlights are special keywords that will notify you across all your devices" (poor English as usual but I'm hoping you get me).

In fact, now that I'm looking at Slack settings, that's what they do: they have multiple tabs, each containing settings that are this-client-only (client) or all-my-clients (server), and they have a lot of helpers next to settings that are not trivial to understand in a couple words.

Either way, I fully agree with you we need splitting, but I think splitting is strictly required before we can achieve fixing the server-settings-not-being-stored-on-server issue or settings sync (this ticket).

> this is where (I think) you are thinking of the lounge incorrectly. Don't compare the lounge to standard irc clients, or a text editor. It's not a traditional client. It's a web app. Compare it to other web apps.

Harsh :D
Of course I don't think of TL as a text editor :) And I usually avoid comparing it to other IRC clients as well. In fact, when I do comparisons, I usually compare with Slack and Telegram, because to me they seem rather close in terms of usage (especially Slack) and technical details (web apps that can be used on mobile - though through native apps, granted).
My link to server-sync for Atom was merely trying to find the closest equivalent for that specific feature. As far as I know, Slack and Telegram do not have that specific feature (probably for the reason that settings in their native apps != settings in their web apps?).

> So, let's take...facebook as an example. Do you ever have settings that are only on 1 web browser for facebook? That would be really weird and confusing.

FB is closer to a SaaS model, so also fairly different. And they don't have client settings at all, as far as I know. I have never used their mobile apps, but I wouldn't be surprised if their settings on mobile only belong to the mobile interface.
In fact, I could be wrong but it seems that all of FB settings in their webapp are account settings, i.e. they belong to your account (server) and not client. In our case, they're are definitely settings I would use on one device but not another (sound, push notifications - on on mobile, off on laptop - themes if someone was working on a high-quality theme for mobile display specifically, etc.). I'm browsing through the FB settings, and I don't see something that I would expect to differ.

> So let's take slack, or discord, or facebook messenger. They generally don't have client settings that are specific to the browser you use.

Not to the browser, but to the client, yes. I've never used Discord or FB messenger so I won't make assumptions, but I have Slack settings that are different on my laptop app and mobile app. For example, by default the laptop app is configured to notify you on `@here`, but not the phone app (there might be subtleties, but I believe that's the case).
Things like 24h format are tricky because they're based on assumptions. For example, Slack server-syncs them, and it makes sense because it's unlikely that you prefer reading time differently based on device lol. But then a "Show seconds" setting (they don't have it, but we do) that would intuitively live next to that 24h format setting, I know that some people like to have it on laptop but not on mobile (to save space)... Tricky.

> As long as we have the feature, that's the main thing.

👍

> It seems to me that most of what you address in the first points should be the focus of a different PR altogether as this issue is focussed on user settings. Heck, before I started on this I had no idea that there were settings that "do not belong" in that regard. I am getting the feeling that you are still compounding to many issues into this one.

Not really though: the reason why I'm bringing that up is because it seems to me that to correctly achieve settings-sync, we need to solve that issue first. Otherwise we end up with the situation you had to go for with the `alwaysSync` thing. And that also means that fixing that issue is now more difficult because it has to go through server-sync and the special case it adds for it.
In a nutshell, I'm just not sure how we can code settings-sync correctly without fixing that issue. But fixing that issue is probably faster than all the time we spent writing these messages lol, it's just that no one ever did it.

> I think that existing users should have it off by default, but new users should have it on by default.

If we were to do that, how would we go about it? Is there a straightforward way we could do that, or would we have to do some extra checks like "does this user have `localStorage` set?"? Also, since the fact that settings-sync option is not stored on the server itself, do we have a way to distinguish a new user vs. existing user on a new device/browser?
It seems to me that if we want to enable this by default, it's extra code and potentially extra configuration, not just a simple boolean switch like our other settings.

**Comment by creesch** [comment]
In response to your last bit, it is relatively straightforward to include something like a flag in the addUser functionality. 

I still think that in your last reply you are compounding too much other things into one issue causing a lot of noise in the communication. I realize you said you wanted to derail the discussion here, but the majority of your reply is about things that imho are outside of the scope of simple setting sync that is a per client opt-in function anyway. 

So again, my proposal is to leave this this issue on the subject of getting sync to work period. I will build it in such a way  (already did in fact)  that is aimed at future expand ability. I think there is a different issue already that is more generally aimed at settings and I think the majority of the discussion in this issue belongs there.  

> Not really though: the reason why I'm bringing that up is because it seems to me that to correctly achieve settings-sync, we need to solve that issue first. Otherwise we end up with the situation you had to go for with the alwaysSync thing. And that also means that fixing that issue is now more difficult because it has to go through server-sync and the special case it adds for it.

This a good example of you compounding in an issue. It is a setting that needs to be synced in all cases. It now will be always synced as it is defined as such in an array, so it is not hardcoded. So whenever this is solved through other means it is simply a case of removing the setting from the `alwaysSync` array. 
I really think you are making issues where there aren't any as of yet and also not will be in the future. 

To be 100% frank with you, settings as I found them are a complete and utter mess in functionality. The client part of them anyway, settings would be written multiple times in various situations where they shouldn't be written and the fact that applying the settings to the input elements worked was actually more of a fluke than by good design. Which is no wonder looking at the history because people kept adding small things without looking at the general shape of settings.

I have already restructured them and moved different functionality to separate functions, etc. The result being that settings functionality is much easier to understand, more robust and generally easier to work with. 
If anything the end result of my work on this issue will be that settings in general are in a much better shape to be looked at and refined. 

So even if the functionality in the syncing department doesn't entirely align with your ideal image, the general shape of settings will be much better and make it easier in the future to work on achieving your ideal as anyone who would want to tackle that would need to do the same amount of refactoring I already did. :)
